x/tokenomics/keeper/settle_session_accounting.go

```diff
@@ -2,15 +2,138 @@ package keeper
+// TODO_BLOCKER: This is just a first naive implementation of the business logic.
 func (k TokenomicsKeeper) SettleSessionAccounting(
 	goCtx context.Context,
 	claim *suppliertypes.Claim,
 ) error {
+	// Parse the context
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	logger := k.Logger(ctx).With("method", "SettleSessionAccounting")
+
+	if claim == nil {
+		logger.Error("received a nil claim")
+		return types.ErrTokenomicsClaimNil
+	}
+
+	// Make sure the session header is not nil
+	sessionHeader := claim.SessionHeader
+	if sessionHeader == nil {
+		logger.Error("received a nil session header")
+		return types.ErrTokenomicsSessionHeaderNil
+	}
+
+	// Validate the session header
+	if err := sessionHeader.ValidateBasic(); err != nil {
+		logger.Error("received an invalid session header", "error", err)
+		return types.ErrTokenomicsSessionHeaderInvalid
+	}
+
+	// Decompose the claim into its constituent parts for readability
+	supplierAddress, err := sdk.AccAddressFromBech32(claim.SupplierAddress)
+	if err != nil {
+		return types.ErrTokenomicsSupplierAddressInvalid
+	}
+	applicationAddress, err := sdk.AccAddressFromBech32(claim.SessionHeader.ApplicationAddress)
+	if err != nil {
+		return types.ErrTokenomicsApplicationAddressInvalid
+	}
+	root := (smt.MerkleRoot)(claim.RootHash)
+
+	// Retrieve the application
+	application, found := k.appKeeper.GetApplication(ctx, applicationAddress.String())
+	if !found {
+		logger.Error(fmt.Sprintf("application for claim with address %s not found", applicationAddress))
+		return types.ErrTokenomicsApplicationNotFound
+	}
+
+	// Retrieve the sum of the root as a proxy into the amount of work done
+	claimComputeUnits := root.Sum()
+
+	logger.Info(fmt.Sprintf("About to start settling claim for %d compute units", claimComputeUnits))
+
+	// Retrieve the existing tokenomics params
+	params := k.GetParams(ctx)
+
+	// Calculate the amount of tokens to mint & burn
+	upokt := sdk.NewInt(int64(claimComputeUnits * params.ComputeUnitsToTokensMultiplier))
+	upoktCoin := sdk.NewCoin("upokt", upokt)
+	upoktCoins := sdk.NewCoins(upoktCoin)
+
+	logger.Info(fmt.Sprintf("%d compute units equate to %d uPOKT for session %s", claimComputeUnits, upokt, sessionHeader.SessionId))
+
+	// NB: We are doing a mint & burn + transfer, instead of a simple transfer
+	// of funds from the supplier to the application in order to enable second
+	// order economic effects with more optionality. This could include funds
+	// going to pnf, delegators, enabling bonuses/rebates, etc...
+
+	// Mint uPOKT to the supplier module account
+	if err := k.bankKeeper.MintCoins(ctx, suppliertypes.ModuleName, upoktCoins); err != nil {
+		return types.ErrTokenomicsApplicationModuleFeeFailed
+	}
+
+	logger.Info(fmt.Sprintf("minted %d uPOKT in the supplier module", upokt))
+
+	// Sent the minted coins to the supplier
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(
+		ctx, suppliertypes.ModuleName, supplierAddress, upoktCoins,
+	); err != nil {
```

Comment:
Just a thought - should we attempt to rollback any minting done before returning an error?

Reply:
This did cross my mind and we definitely need to do if it's not atomic.

Adding the following comment in the meantime to block merging

```
// TODO_IN_THIS_PR: Rollback state changes if we find out that this function is not
// atomic.if this function is not atomic.
```

---

CCing @tac0turtle - Do you know at what scope / granularity keeper functions are atomic? I.e. if an error is returned, do the state changes get rolled back?

Reply:
Thanks Marko!

![Screenshot 2024-01-26 at 9 16 22 AM](https://github.com/pokt-network/poktroll/assets/1892194/af6c4e11-5ca1-41e0-82a1-b4f8051a3696)